options.go

```diff
@@ -41,684 +28,118 @@ func (o Options) Add(key OptionCode, value encoding.BinaryMarshaler) error {
-//
-// The VendorClass structure returned contains VendorClass in
-// the option.
-//
-// The boolean return value indicates if OptionVendorClass was present in the
-// Options map.  The error return value indicates if any errors were present
-// in the VendorClass data.
-func (o Options) VendorClass() (*VendorClass, bool, error) {
-	v, ok := o.Get(OptionVendorClass)
-	if !ok {
-		return nil, false, nil
-	}
-
-	vc := new(VendorClass)
-	err := vc.UnmarshalBinary(v)
-	return vc, true, err
-}
-
-// VendorOpts returns the Vendor-specific Information Option value, described in RFC 3315,
-// Section 22.17.
-//
-// The VendorOpts structure returned contains Vendor-specific Information data present in
-// the option.
-//
-// The boolean return value indicates if VendorOpts was present in the
-// Options map.  The error return value indicates if any errors were present
-// in the class data.
-func (o Options) VendorOpts() (*VendorOpts, bool, error) {
-	v, ok := o.Get(OptionVendorOpts)
-	if !ok {
-		return nil, false, nil
+// GetOne works just like Get, but if there is more than one value for the
+// OptionCode key, ErrInvalidPacket will be returned.
+func (o Options) GetOne(key OptionCode) ([]byte, error) {
+	vv, err := o.Get(key)
+	if err != nil {
+		return nil, err
 	}
 
-	vo := new(VendorOpts)
-	err := vo.UnmarshalBinary(v)
-	return vo, true, err
-}
-
-// InterfaceID returns the Interface-Id Option value, described in RFC 3315,
-// Section 22.18.
-//
-// The InterfaceID structure returned contains any raw class data present in
-// the option.
-//
-// The boolean return value indicates if InterfaceID was present in the
-// Options map.  The error return value indicates if any errors were present
-// in the interface-id data.
-func (o Options) InterfaceID() (InterfaceID, bool, error) {
-	v, ok := o.Get(OptionInterfaceID)
-	if !ok {
-		return nil, false, nil
+	if len(vv) != 1 {
+		return nil, ErrInvalidPacket
 	}
-
-	var i InterfaceID
-	err := i.UnmarshalBinary(v)
-	return i, true, err
+	return vv[0], nil
 }
 
-// IAPD returns the Identity Association for Prefix Delegation Option value,
-// described in RFC 3633, Section 9.
-//
-// Multiple IAPD values may be present in a a single DHCP request.
-//
-// The boolean return value indicates if OptionIAPD was present in the Options
-// map.  The error return value indicates if one or more valid IAPDs could not
-// be parsed from the option.
-func (o Options) IAPD() ([]*IAPD, bool, error) {
-	// Client may send multiple IAPD option requests, so we must
-	// access the map directly
-	vv, ok := o[OptionIAPD]
-	if !ok {
-		return nil, false, nil
-	}
-
-	// Parse each IA_PD value
-	iapd := make([]*IAPD, len(vv))
-	for i := range vv {
-		ia := new(IAPD)
-		if err := ia.UnmarshalBinary(vv[i]); err != nil {
-			return nil, true, err
-		}
-
-		iapd[i] = ia
-	}
-
-	return iapd, true, nil
-}
+// Marshal writes options into the provided Buffer sorted by option codes.
+func (o Options) Marshal(b *util.Buffer) {
```

Comment:
Is there any way we can not expose the buffer type?  I feel like it belongs in an internal package, and that this function should just have the usual `MarshalBinary() ([]byte, error)` signature.

Reply:
This is similar to the optslice.write([]byte) function you had previously -- it avoids having to allocate all options in one buffer and then copying that buffer to the "parent" buffer.

The only reason this has to be exported is because of the usage in dhcp6opts. Otherwise, even Marshal and Unmarshal could be implementation details of this package.

(I'll push a commit to change this to what you want, but I'd prefer to revert it.)